eos-upgrade-eos2-to-eos3

```diff
@@ -0,0 +1,222 @@
+    if [ ${major_version} == 2 ]; then
+	echo "Please update the OS and try again"
+    elif [ ${major_version} == 3 ]; then
+	echo "You already are running Endless OS version 3"
+    else
+	echo "You are running an unsupported version: ${version}"
+    fi
+    exit 1
+fi
+
+if ! $FORCE; then
+    echo
+    echo "You are about to upgrade Endless OS from version 2 to version 3."
+    echo
+    echo "Endless OS 3 uses a new technology for applications called flatpak,"
+    echo "which is a standard format for deploying Linux applications that enables"
+    echo "Endless to run applications distributed by other development teams."
+    echo
+    echo "The legacy application format is no longer supported, and the first step"
+    echo "of the upgrade process is to remove all installed application bundles."
+    echo "Then, roughly a gigabyte of data will be downloaded from the internet,"
+    echo "which may take a long time and incur data charges depending on your"
+    echo "internet connection and contract with your provider."
+    echo
+    echo "In the meantime, while downloading Endless OS 3, you may continue"
+    echo "to use Endless OS 2 and some of the core applications such as"
+    echo "the internet browser and office applications."
+    echo "Once the upgrade to Endless OS 3 is downloaded and ready, you will be"
+    echo "asked to reboot, and then you will be able to install new flatpak apps"
+    echo "from the Endless OS 3 app center."
+    echo
+    echo "Once the download completes, it will take several additional minutes"
+    echo "for the deployment to complete -- please be patient and wait until"
+    echo "you see the 'Upgrade complete!' message."
+    echo
+    read -p "Are you sure you want to remove all legacy app bundles and proceed? [y/N] "
+    response=${REPLY,,} # to lower
+    if [[ ! $response =~ ^(yes|y)$ ]]; then
+        exit 1
+    fi
+fi
+
+# Stop any updates in progress
+killall ostree-daemon || true
+killall eos-updater || true
+
+# For older installs (eos2.3.x and earlier), we had a separate boot
+# partition which we no longer support on eos3
+if [ ! "$(ls -A /sysroot/boot)" ]; then
+    echo "Migrating separate boot partition into main partition"
+    # Get the boot partition (e.g., /dev/sda1 or /dev/mmcblk0p1)
+    boot_part=`mount | grep /boot | awk '{print $1}'`
+    # Extract the device name (e.g., /dev/sda or /dev/mmblk0)
+    boot_dev=`echo $boot_part | sed 's/p*[0-9]\+$//'`
+    # Copy the contents from /boot to the main partition
+    cp -a /boot/* /sysroot/boot
+    # Manually unmount the boot partition
+    umount /boot
+    # Re-install the bootloader, telling it to look for config files
+    # in the main partition
+    grub-install --boot-directory=/sysroot/boot --modules="ext2 part_msdos part_gpt" --directory=/usr/lib/grub/i386-pc $boot_dev
+    sync
+    # Remove the old boot partition
+    sfdisk --dump $boot_dev | grep -v $boot_part | sfdisk --force $boot_dev
```

Comment:
I saw behaviour in the past with sfdisk opening the device as read-write in similar situations. This causes a race where upon close, udev recreates all the partition devices meaning that the followup sfdisk command will (occasionally) fail. So please do check this or at least send me the strace log from 2.6.10 and I'll look.